modules/lang-painless/src/main/java/org/elasticsearch/painless/Def.java

```diff
@@ -1268,4 +1277,30 @@ static MethodHandle arrayIndexNormalizer(Class<?> arrayType) {
 
         private ArrayIndexNormalizeHelper() {}
     }
+
+    public static class Encoding {
```

Comment:
Do you think it'd make the def handling code easier to read if you made a parsing ctor and used this object instead? It'd also let you assert the round trip and stuff.

Reply:
It could have actual methods on it to build the signature of the factory method and stuff.

Reply:
Agreed.

Reply:
It's a decent amount of code but it does increase the likelihood of us finding a surprise early.  What do you think?

Reply:
I talked with @stu-elastic on this one and he decided not to do it in this PR - it'd just be too big and we don't have enough coverage to make sure we don't accidentally break something. It's still a good change, but a follow up one.